src/OrbitBase/include/OrbitBase/CanceledOr.h

```diff
@@ -17,30 +18,45 @@ struct Canceled {};
 
 // CanceledOr can be used as the return type of an cancelable operation. Based on std::variant.
 // Check whether CanceledOr object is canceled, use orbit_base::IsCanceled. Get the value of a non
-// canceled CanceledOr with std::get<T>().
+// canceled CanceledOr with the `GetValue` member function or the `GetNotCanceled` free function.
+//
+// `CanceledOr<T>` is implemented by privately deriving from `std::variant`. This allows us to take
+// advantage of std::variant's carefully crafted constructors while still keeping std::variant an
+// implementation detail. Implementing those constructors by hand would add a lot of complexity.
 template <typename T>
-using CanceledOr = std::variant<VoidToMonostate_t<T>, Canceled>;
+class CanceledOr : private std::variant<VoidToMonostate_t<T>, Canceled> {
+ public:
+  using Value = VoidToMonostate_t<T>;
 
-// Free function to quickly check whether a CanceledOr type is canceled. Abstracts
-// std::holds_alternative
+  [[nodiscard]] bool HasValue() const { return this->index() == 0; }
+  [[nodiscard]] bool IsCanceled() const { return !HasValue(); }
+
+  [[nodiscard]] const Value& GetValue() const& { return std::get<0>(*this); }
+  [[nodiscard]] const Value& GetValue() const&& { return std::get<0>(*this); }
+  [[nodiscard]] Value GetValue() && { return std::get<0>(std::move(*this)); }
```

Comment:
`get( /*const*/tuple<Types...>&&)` returns by `/*const*/ &&`. Why do we return by value and lvalue ref here?

Reply:
Yes, makes sense. Let me quickly fix that.

Reply:
Done.